core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -800,6 +832,78 @@ private[spark] class TaskSetManager(
     maybeFinishTaskSet()
   }
 
+  private[scheduler] def updateBlacklistForFailedTask(
+      host: String,
+      exec: String,
+      index: Int): Unit = {
+    val failureStatus = execToFailures.getOrElseUpdate(exec, new FailureStatus(host))
+    failureStatus.totalFailures += 1
+    failureStatus.tasksWithFailures += index
+
+    // check if this task has also failed on other executors on the same host, and if so, blacklist
+    // this task from the host
+    val execsWithFailuresOnNode = nodesToExecsWithFailures.getOrElseUpdate(host, new HashSet())
+    execsWithFailuresOnNode += exec
+    val failuresOnHost = (for {
+      exec <- execsWithFailuresOnNode.toIterator
+      failures <- execToFailures.get(exec)
+    } yield {
+      if (failures.tasksWithFailures.contains(index)) 1 else 0
+    }).sum
+    if (failuresOnHost >= MAX_TASK_ATTEMPTS_PER_NODE) {
+      nodeBlacklistedTasks.getOrElseUpdate(host, new HashSet()) += index
+    }
+
+    if (failureStatus.totalFailures >= MAX_FAILURES_PER_EXEC_STAGE) {
```

Comment:
This logic seems to be repeated in at least one other place (where you first blacklist something on an executor, and then escalate the issue to the host). I wonder if there's a helper class that could be added that would handle this in both cases (i.e., for a specific task, and for a whole stage, and I think also for the app)? Maybe there's not enough shared logic?